include/common/array-size.h

```diff
@@ -0,0 +1,28 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+#ifndef LABWC_ARRAY_SIZE_H
+#define LABWC_ARRAY_SIZE_H
+
+/*
+ * These macros were copied from the Linux Kernel because its ARRAY_SIZE() macro
+ * breaks the compile if anyone tries to use it on a pointer.
+ *
+ * https://github.com/torvalds/linux/blob/master/include/linux/
+ *
+ * Both gcc and clang support __builtin_types_compatible_p()
+ *
+ */
+
+#define __same_type(a, b) __builtin_types_compatible_p(__typeof(a), __typeof(b))
+
+#define BUILD_BUG_ON_ZERO(e) (sizeof(struct { int: -!!(e); }))
+
+/* &a[0] degrades to a pointer: a different type from an array */
+#define __must_be_array(a) BUILD_BUG_ON_ZERO(__same_type((a), &(a)[0]))
+
+/**
+ * ARRAY_SIZE - get the number of elements in array @arr
+ * @arr: array to be sized
+ */
+#define ARRAY_SIZE(arr) (sizeof(arr) / sizeof((arr)[0]) + __must_be_array(arr))
```

Comment:
This is a bit much undocumented dark magic for my taste:

So `__builtin_types_compatible_p()` appears to return `1` if the types are the same: https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html#index-_005f_005fbuiltin_005ftypes_005fcompatible_005fp

I have absolutely no idea what happens in `BUILD_BUG_ON_ZERO`. Couldn't we just do an `static_assert(__same_type((a), &(a)[0]) == 1, "Tried to use ARRAY_SIZE on a non-array")` or something?

Finally `ARRAY_SIZE()`  itself, I don't understand why it adds the result of `__must_be_array()`, if the naming of the macros (e.g. "build-bug on zero") is correct, shouldn't it always return something `> 0` here and thus return the wrong array size?

I assume somehow those macros make sense and if that is the case I think it's fine using them but we should at least document the way they work properly. Otherwise stupid people like me end up reading the code and just thinking.. "no clue what I've just read, guess its fine".

Reply:
BUILD_BUG_ON_ZERO is basically static_assert().

Reply:
```
#define ARRAY_SIZE(a) { \
	static_assert(__builtin_types_compatible_p(typeof(a), typeof(&(a)[0])) == 0); \
	(sizeof(a) / sizeof(a[0])); \
}
```

Reply:
I'm really sorry for OCD in the comments, I know you receive an email for every change I make >_<

Reply:
> ```
> #define same_type(a, b) (__builtin_types_compatible_p(typeof(a), typeof(b)) == 0)
> #define ARRAY_SIZE(a) ({ static_assert(same_type(a, &(a)[0])); sizeof(a) / sizeof(a[0]); })
> ```

`same_type()` in your example is either misnamed  or wrong as it returns the types being different rather than being same / compatible.

I'd also use the 2nd variant of `static_assert(expr, message)` and thus throw in a friendly error message that shows up during compile if the `static_assert()` fails.

But other than that, this two line variant indeed looks way easier to understand to me.

Reply:
Btw, I am not sure if we actually need the extra check here, at least on my setup we already get an error (with `--werror`) without the check:
```
../src/main.c:82:34: error: division ‘sizeof (char *) / sizeof (char)’ does not compute the number of array elements [-Werror=sizeof-pointer-div]
   82 | #define ARRAY_SIZE(a) (sizeof(a) / sizeof((a)[0]))
      |                                  ^
../src/main.c:90:12: note: in expansion of macro ‘ARRAY_SIZE’
   90 |  int len = ARRAY_SIZE(bar);
      |            ^~~~~~~~~~
../src/main.c:89:8: note: first ‘sizeof’ operand was declared here
   89 |  char *bar = malloc(100)
```

Reply:
Nope, it returns zero on a match.

Reply:
```
#define same_type(a, b) (__builtin_types_compatible_p(typeof(a), typeof(b)) == 0)
````
The name is off of course. `static_assert()` should fail exactly when `typeof(a) == typeof(&a[0])`. I should really start to write code in the night.